### PR TITLE
Fix font clipping and harfbuzz hyphenation rendering issues

### PR DESCRIPTION
**Pull request by poire-z**
Fix some of the issues noticed and discussed along https://github.com/koreader/koreader/issues/4577.
Including the rendering issue with harfbuzz kerning full (aka "best") described in https://github.com/koreader/koreader/issues/4577#issuecomment-462908911.
See individual commit messages for detail.

### CODE REVIEW

File: crengine/src/lvtextfm.cpp [hunk 1]
```diff
@@ -1170,7 +1194,32 @@ class LVFormatter {
...
+                        int new_width;
+                        if ( measureWord(word, new_width) ) {
+                            word->width = new_width;
+                            word->min_width = word->width;
+                        }
+                    }
                     if ( m_flags[i-1] & LCHAR_ALLOW_HYPH_WRAP_AFTER ) {
+                        if ( m_flags[i] & LCHAR_IS_LIGATURE_TAIL ) {
+                            // The end of this word is part of a ligature that, because
+                            // of hyphenation, has been splitted onto next word.
```
Comment (Frenzie) [inline crengine/src/lvtextfm.cpp]: was split onto the next line?

(past tense of split is split but my comment is about word vs line)

**Comment by Frenzie** [comment]
Oh btw, don't you need one of those cache version updates for the "best" ligature fix?

File: crengine/src/lvtextfm.cpp (hunk 1 above)
Comment (poire-z) [reply crengine/src/lvtextfm.cpp]: the ligature is split: `zzzffzzz` split as `zzzf` and `fzzz`. Each will indeed be on different lines, but each part makes a "word" in the context of lvtextfm (a real word hyphenated is 2 "word" objects)

Comment (Frenzie) [reply crengine/src/lvtextfm.cpp]: Yeah, that crossed my mind just after pressing "comment" but I figured I'd leave it to be completely sure. :-)

**Comment by poire-z** [comment]
I don't think so, why ?
I'll work out this cache thing and huge font size issues in a next PR, but I think this HB fix does not need any of that yet. Pushing these harmless (I hope) commits independantly first :)

**Comment by Frenzie** [comment]
It's only needed for different DOM trees, not for completely upending the number of lines? Alright, nice.

**Comment by poire-z** [comment]
Oh, right ! `FORMATTING_VERSION_ID` ! Yes indeed !
(Thought you were talking about the glyph width or lsb/rsb cache I had in mind in that other long issue:)

**Comment by poire-z** [comment]
Just adding a note for future boring investigation:
I still got that same rendering issue (words overlapping, we got with HB "best") with some italic words and some Baskerville fonts (which have kerning for `ff`), with both "good" and "best", with
```
blahblah blahblah ef-
froyablement ennuyeux  (spacing between these words was gone)
```

**Comment by poire-z** [comment]
Mentionning another issue I just noticed (because I recently enjoyed reading in 2 columns, and because of the smaller lines width, some issues become more noticable :/):
image
Because of the souf-flai hyphenation at a point that would be a `ff` ligature, the second line get its starting `f` with a glyph width of 0, which allows the word "mon" to be included as fitting on that second line - but when actually adding this line, as it does not really fit, we end up reducing the widths of all spaces to have it fit.
(Probably the same issue I mentionned in the post just above?)

```diff
--- a/crengine/src/lvtextfm.cpp
+++ b/crengine/src/lvtextfm.cpp
@@ -4595,6 +4595,22 @@ public:
                 maxWidth = getCurrentLineWidth();
             }

+            if ( m_flags[pos] & LCHAR_IS_CLUSTER_TAIL ) {
+                // This line start with a cluster tail, probably because hyphenation was
+                // allowed inside this cluster. The first char(s) would get a width of 0,
+                // which may allow more text to be brought into this line: later, in AddLine(),
+                // we may have to handle the excess of text by reducing all spaces' widths
+                // and possibly making them all 0 if needed.
+                // So, account for the whole cluster width into this line (by considering it as a
+                // negative spaceReduceWidth): it might be too much and we could do with a fraction
+                // of it (but which value?), but better too much spacing than not enough.
+                int bpos = pos - 1;
+                while ( bpos > 0 && m_flags[bpos] & LCHAR_IS_CLUSTER_TAIL )
+                    bpos--;
+                int cluster_width = (m_widths[bpos] - (bpos > 0 ? m_widths[bpos-1] : 0));
+                spaceReduceWidth -= cluster_width;
+            }
+
             // Find candidates where end of line is possible
             bool seen_non_collapsed_space = false;
             bool seen_first_rendered_char = false;
```
will allow getting:
image

**Pull request merged**